text/001x-battery_improvements.md

```diff
@@ -0,0 +1,294 @@
+- New `battery_status` (`MAV_BATTERY_STATUS_FLAGS`) replaces [`fault_bitmask`](https://mavlink.io/en/messages/common.html#MAV_BATTERY_FAULT), [`charge_state`](https://mavlink.io/en/messages/common.html#MAV_BATTERY_CHARGE_STATE) and [`mode`](https://mavlink.io/en/messages/common.html#MAV_BATTERY_MODE)
+  ```xml
+  <field type="uint8_t" name="charge_state" enum="MAV_BATTERY_CHARGE_STATE">State for extent of discharge, provided by autopilot for warning or external reactions</field>
+  <field type="uint8_t" name="mode" enum="MAV_BATTERY_MODE">Battery mode. Default (0) is that battery mode reporting is not supported or battery is in normal-use mode.</field>
+  <field type="uint32_t" name="fault_bitmask" display="bitmask" enum="MAV_BATTERY_FAULT">Fault/health/ready-to-use indications.</field>
+  ```
+  - The state is almost all inferred from the battery remaining. Traditionally the GCS sets appropriate levels for warning - the battery has a less accurate view on appropriate warning levels for a particular system, making this redundant. The exception is charging status.
+  - The mode and fault are reasonable, but make more sense as general status indication.
+  - The status loses some faults which make no sense (e.g. `MAV_BATTERY_FAULT_SPIKES`) or which can now be inferred (like deep discharge).
+
+
+#### Questions
+
+- Do we need all the other fields?
+- Are there any other fields missing?
+- Should we have `MAV_BATTERY_FAULT` for critical level "just before deep discharge"? The battery can know this, and it might prevent over discharge.
+  ```xml
+      <entry value="1024" name="BATTERY_FAULT_CRITICAL_LEVEL">
+        <description>Battery is at critically low level. Undamaged, but not ready to use.</description>
+      </entry>
+  ```
+- `MAV_BATTERY_STATUS_FLAGS`
+  - Should this include `MAV_BATTERY_STATUS_FLAGS_IN_USE` (connected, drawing significant power to fly). Not clear when you would need this or how you would use it. Or precisely what would count as "in use".
+
+## Battery Cell Voltages
+
+The proposed battery message is:
+
+```xml
+    <message id="371" name="BATTERY_CELL_VOLTAGES">
+      <description>Battery cell voltages.
+        This message is provided primarily for cell fault debugging.
+        For batteries with more than 12 cells the message should be sent multiple times, iterating the index value.
+        It should be streamed at very low rate (less than once a minute) or streamed only on request.</description>
+      <field type="uint8_t" name="id" instance="true">Battery ID</field>
+      <field type="uint8_t" name="index">Cell index (0 by default). This can be iterated for batteries with more than 12 cells.</field>
+      <field type="uint16_t[12]" name="voltages" units="mV" invalid="[0]">Battery voltage of 12 cells at current index. Cells above the valid cell count for this battery should be set to 0.</field>
+    </message>
```

Comment:
@julianoes @auturgy We make this a short fixed length of 12 cells, because we wanted to make sure that it would be smaller in the general case, and in particular if extended.

HOwever it occurs to me that
1. Given the name, we should never extend this message.
2. If we make 0 the default and extend to 126 bytes then just as efficient due to zero bye truncation for short messages. But MORE efficient for bigger than that because we don't add packet overheads for each index.
3. Much easier for consumers to use, up to 126 cells. 
4. Two bytes bigger for the most common case - up to 12 cells.
5. But this is only send for debugging, and realistically you're not using it for "control" but for long term trends. You could probably get a way with a ping every few minutes, or even hours.

So I'm suggesting we finalise the message prioritising easy of use. Thoughts?

```suggestion
    <message id="371" name="BATTERY_CELL_VOLTAGES">
      <description>Battery cell voltages.
        This message is provided primarily for cell fault debugging.
        For batteries with more than 12 cells the message should be sent multiple times, iterating the index value.
        It should be streamed at very low rate (less than once a minute) or streamed only on request.</description>
      <field type="uint16_t" name="id" instance="true">Battery ID</field>
      <field type="uint16_t" name="index">Cell index (0 by default). This can be iterated for batteries with more than 12 cells.</field>
      <field type="uint16_t[126]" name="voltages" units="mV" invalid="[0]">Battery voltage of up to 126 cells at current index. Cells above the valid cell count for this battery should be set to 0.</field>
    </message>
```

Reply:
Merging. We can discuss the detail later.

Reply:
@dakejahl FYI

Reply:
LGTM

Reply:
I think I'm adding wrong or not understanding something? 
126*2 (cells) + 2 (id) + 2 (instance) = 256 B > payload[max 255]